src/main/java/graphql/util/Traverser.java

```diff
@@ -164,6 +164,7 @@ public void reset() {
             } else {
                 TraversalControl traversalControl = visitor.enter(currentContext);
                 assertNotNull(traversalControl, "result of enter must not be null");
+                this.traverserState.addVisited((T) currentContext.thisNode());
```

Comment:
In the original version this was a side effect of `currentContext.isVisited()` call.
The reason was to avoid redundant lookup in the table with the same key. With this version this redundancy happens when `isVisited` calls `visited.contansKey()` and here when `addVisited` calls `visited.put()`.

Reply:
true, but having a side effect in isVisited I consider not ok: it is extremely unexpected and makes understanding the code very hard. And I don't mind the double lookup, because I don't think it has a real measurable performance impact.